lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1647,7 +1681,12 @@ public function remove_attribute( $name ) {
 		 * @see https://html.spec.whatwg.org/multipage/syntax.html#attributes-2:ascii-case-insensitive
 		 */
 		$name = strtolower( $name );
-		if ( $this->is_closing_tag || ! isset( $this->attributes[ $name ] ) ) {
+
+		if ( ! isset( $this->attributes[ $name ] ) ) {
+			if ( isset( $this->lexical_updates[ $name ] ) ) {
+				unset( $this->lexical_updates[ $name ] );
+			}
+
```

Comment:
I don't think it would process this case correctly?

```php
$p = new WP_HTML_Tag_Processor('<div id="a"></div>');
$p->next();
$p->set_attribute('id', 'b');
$p->remove_attribute('id');
```

How about changing the condition like this?
```suggestion
		if ( isset( $this->lexical_updates[ $name ] ) ) {
			unset( $this->lexical_updates[ $name ] );
		}
			
		if ( ! isset( $this->attributes[ $name ] ) ) {

```

This way any enqueued updates will be removed regardless of whether the attribute already existed.

Reply:
Also should this method call `class_name_updates_to_attributes_updates()` now?

Reply:
> I don't think it would process this case correctly?

No, that should work! 😄 

<details>
<summary>
You can try this patch
</summary>

```diff
diff --git a/phpunit/html/wp-html-tag-processor-test.php b/phpunit/html/wp-html-tag-processor-test.php
index 662749a523..55e5ba8c24 100644
--- a/phpunit/html/wp-html-tag-processor-test.php
+++ b/phpunit/html/wp-html-tag-processor-test.php
@@ -605,6 +605,23 @@ class WP_HTML_Tag_Processor_Test extends WP_UnitTestCase {
                $this->assertSame( self::HTML_SIMPLE, $p->get_updated_html() );
        }
 
+       /**
+        * @ticket 56299
+        *
+        * @covers set_attribute
+        * @covers remove_attribute
+        * @covers get_attribute
+        * @covers get_updated_html
+        */
+       public function test_get_attribute_reflects_set_attribute_followed_by_remove_attribute_for_existing_attribute_before_it_is_updated() {
+               $p = new WP_HTML_Tag_Processor( self::HTML_SIMPLE );
+               $p->next_tag();
+               $p->set_attribute( 'id', 'test-value' );
+               $p->remove_attribute( 'id' );
+               $this->assertNull( $p->get_attribute( 'id' ) );
+               $this->assertSame( '<div ><span id="second">Text</span></div>', $p->get_updated_html() );
+       }
+
        /**
         * @ticket 56299
         *
```

</details>

(I'll push a commit with that test.)

The idea is as follows:
- If the attribute is not set in the HTML, we remove any update that would add it.
- If the attribute _is_ set in the HTML, we override any updates with a new one that will remove it, when applied. That latter part is handled by existing code a couple lines further down (that's not touched here): https://github.com/WordPress/gutenberg/blob/340b51d44558bf97402c46150887d88ff796404c/lib/experimental/html/class-wp-html-tag-processor.php#L1654-L1669

Reply:
> Also should this method call `class_name_updates_to_attributes_updates()` now?

Ah, good point! I think it should indeed 🤔 I'll also add some test coverage !

Reply:
Test coverage in d739f31d14.

I think we can actually get away without `class_name_updates_to_attributes_updates` -- clearing `this$->classname_updates` should be enough: 06245b3726